sdk/cxx/include/sawtooth_sdk.h

```diff
@@ -0,0 +1,258 @@
+
+// The transaction data for a Transaction Processing request.
+class Transaction final {
+ public:
+    Transaction(TransactionHeaderPtr header, StringPtr payload, StringPtr signature):
+            header_(header), payload_(payload), signature_(signature) {
+    }
+
+    Transaction (const Transaction&) = delete;
+    Transaction (const Transaction&&) = delete;
+    Transaction& operator= (const Transaction&) = delete;
+
+    const TransactionHeaderPtr header() const {
+        return this->header_;
+    }
+
+    const std::string& payload() const {
+        return *(this->payload_);
+    }
+
+    const std::string& signature() const {
+        return *(this->signature_);
+    }
+
+ private:
+    TransactionHeaderPtr header_;
+    StringPtr payload_;
+    StringPtr signature_;
+};
+typedef std::unique_ptr<Transaction> TransactionUPtr;
+
+
+// Provides access to the global state
+class GlobalStateIF {
+ public:
+    typedef std::pair<std::string, std::string> KeyValue;
+
+    virtual ~GlobalStateIF() {}
+
+    // Retrieve a single value from global state. If you are retrieving
+    // multiple values, it is encouraged to use the batch get overload defined
+    // below. The function returns false if the value is not retrieved, true if
+    // it is found in the GlobalStore, and throws an exception if an error occurs
+    virtual bool GetState(std::string* out_value, const std::string& address) const = 0;
+
+    // Retrieve multiple values from the global state. Values are returned in
+    // the out_values map. If a value is not present then there will be no
+    // entry in the returned map. So you must check for the presence of a
+    // returned value prior to accessing it.
+    virtual void GetState(std::unordered_map<std::string, std::string>* out_values,
+        const std::vector<std::string>& addresses) const = 0;
+
+    // Set a single GlobalState value. Use the batch Set function defined
+    // below if you are setting multiple items.
+    virtual void SetState(const std::string& address, const std::string& value) const = 0;
+
+    // Set multiple values in the global state. Each entry in the addresses
+    // param is a std::pair, with the first value representing the address and
+    // the second the value.
+    //      std::vector<GlobalState::KeyValue> addresses;
+    //      addresses.push(std::make_pair(address, value));
+    virtual void SetState(const std::vector<KeyValue>& addresses) const = 0;
+
+    // Delete a single GlobalState value. Used the batch Delete function
+    // defined below if you are deleting multiple items.
+    virtual void DeleteState(const std::string& address) const = 0;
+
+    // Delete multiple entries from global state.
+    virtual void DeleteState(const std::vector<std::string>& address) const = 0;
+};
+typedef std::shared_ptr<GlobalStateIF> GlobalStatePtr;
+typedef std::unique_ptr<GlobalStateIF> GlobalStateUPtr;
+
+
+// TransactionApplicator is an interface that defined to represent the
```

Comment:
This class is inconsistent with the other SDKs.  It seems to largely duplicate the purpose of TransactionHandler.

Reply:
It is a good point, but it was taken as is from the existing code as is. Its redesign was not in the scope of this change

Reply:
The transaction handler is a singleton and the CXX sdk allows for multithread processing of transactions. It is not good practice to have multiple threads calling into a function on the same object. The transaction applicator is a transaction specific instance of a processor that can be dispatched to a thread for processing with an explicit lifespan of the processing of that transaction. Again this feedback should have been givin on the initial commit.